### PR TITLE
[sysinit] new component

### PR DESCRIPTION
**Pull request by gares**
This PR introduces a new component called `sysinit/` that is in charge of command line argument parsing and initialization for all components below it (all but `stm/` and `toplevel/`). The main entry point and documentation is `sysinit/coqinit.mli`.

In order to make this work, I had to clean up argument parsing, making it more functional (some commands were setting global options/refs on the spot, making it hard to parse arguments when one wants w.r.t. the initialization of the system, like initializing the summary/goptions). There are still many global refs set in `sysinit/coqargs.ml` but none of them interferes with the system initialization. The vast majority of the options in `sysinit/coqargs.ml` are of general interest, I moved some of the `coqtop` specific ones in `toplevel` (eg `-color`) but some of them are still there (eg `-print-emacs`) and are left for future work.

 The `toplevel/` component does depends on the `stm/` and can hence be used to build top levels that use that document manager (like `coqtop`, `coqc`, `coqidetop`). If one wants to build a Coq "process" which does not use the `stm/` can still parse command line options and initialize the system using the new `sysinit/coqinit.ml` module (reusing the code of the new component).

Tested vscoqtop, it runs fine: https://github.com/gares/coq/commit/ed90353211d31bac67c15a60d41ec9849504cc36

### CODE REVIEW

**Comment by gares** [comment]
CC @maximedenes

**Comment by ejgallego** [comment]
From a very quick look I'm not sure this is the way to go; for example all of jscoq, serapi, and coq-lsp, implement their own toplevel and the machinery that should be shared is much smaller than the one touched here. In general you want just a bit of the init code and arguments. I think abstracting as this patch does is anyways moot due to the different needs in terms of execution for each toplevel.

**Comment by gares** [comment]
Did you find another way to share the code of `coq_toplevel.ml` and `coqargs.ml` or you did copy paste it?
Do these toplevels set up VM, native, indices-matter... as coqtop does, or just don't support these features?

**Comment by gares** [comment]
FTR, this is the dune file I have in the vscoq branch
```
(library
 (name toplevel)
 (public_name coq.toplevel)
 (synopsis "Coq's Interactive Shell [terminal-based]")
 (wrapped false)
 (libraries coq.stm toplevel_core)
 (modules coqc coqcargs ccompile coqtop coqloop workerLoop vernac g_toplevel)
 )
; Interp provides the `zarith` library to plugins, we could also use
; -linkall in the plugins file, to be discussed.

(library
 (name toplevel_core)
 (public_name coq.toplevel_core)
 (synopsis "Coq's toplevel")
 (wrapped false)
 (libraries coq.vernac)
 (modules coqinit coqargs coq_toplevel usage)
 )

(coq.pp (modules g_toplevel))
```

**Comment by gares** [comment]
These are the shared modules: `coqinit coqargs coq_toplevel usage` and you can't use them without linking the STM

**Comment by ejgallego** [comment]
> Did you find another way to share the code of `coq_toplevel.ml` and `coqargs.ml` or you did copy paste it?
> Do these toplevels set up VM, native, indices-matter... XDG/rcfile, envars... as coqtop does, or just don't support these features?

Yup they do, as of today they copy / paste, but the total is very few lines of code.

I was thinking of just moving that part to a new library `coq.init`, in fact I already have a patch I didn't manage to finish, as indeed de-synchronization due to cut and paste has caused a few issues in the past for example in SerAPI.

**Comment by ejgallego** [comment]
> These are the shared modules: `coqinit coqargs coq_toplevel usage` and you can use them without linking the STM with this PR

Yeah, I think these modules should be already decouple from the stm in `master`, they need a bit of work for example, coqinit is pretty badly named. Quite a few of code that is in coqtop.ml should be moved there, the current API is still pretty bad [I mean the one in master]

**Comment by gares** [comment]
Concretely, what do you suggest

**Comment by ejgallego** [comment]
> Concretely, what do you suggest

I think an easier path would be to introduce `coq.sysinit` which doesn't depend on the stm, at least that'd remove all duplication in toplevels and I think would improve the current API. Then you could write your toplevel in a more flexible way IMO.

I can submit my tree as a draft PR if you wanna have a look.

**Comment by gares** [comment]
> I can submit my tree as a draft PR if you wanna have a look.

yes, because adding a new component handling initialization (and hence argument parsing) is what this PR tries to do, so I'm not so sure to understand what you propose (other than removing `custom_toplevel` while providing the code it factors as an API).

**Comment by gares** [comment]
> > I can submit my tree as a draft PR if you wanna have a look.
> 
> yes,

yes please

**Comment by gares** [comment]
I pushed a completely different implementation, I'll update the PR description

File: sysinit/dune
Comment (Zimmi48) [inline sysinit/dune:7]: Trivial nit but the usual formatting of Dune files is:
```suggestion
 (libraries coq.vernac))
```

**Review by Zimmi48 — APPROVED** [review]

**Comment by SkySkimmer** [comment]
Why's the classifier moved to vernac, is it used by sysinit somehow?

**Comment by gares** [comment]
My main motivation for this PR is to be able to develop vscoqtop out of tree without linking the STM (but still sharing argument parsing and initialization).
The new document manager still classifies sentences in order to know what they do w.r.t. proof opening/closing and parsing effect.

Even more importantly, this is where that facility belongs (even the file name suggests that):
- The classification of vernacular commands is not in the STM only, in the sense that the syntax in VERNAC EXTEND is not implemented in the STM (if you don't link the stm, you still need the code generated by the macros to make sense)
- Finally there is already a similar classification in vernac used to translate the AST into typed closures. Having them in the same place may help merge the two (both in the syntax and in the information they provide).

**Comment by gares** [comment]
Anyone picking this PR up?

**Comment by gares** [comment]
I'd like this one not to be merged soonish because I need this to move vscoqtop out of tree and because this is tricky code so I'd rather have it tested from now.

Anyone willing to assign?

**Review by SkySkimmer — APPROVED** [review]
I didn't look very indepth as this seems to be mostly code movement, looks OK.

I intend to merge later this week so anyone with comments should speak up (it's OK to ask for a delay).
Special cc to @ejgallego since he's in the toplevel team which is codeowner of the new folder.

File: plugins/ltac/profile_ltac.ml
Comment (SkySkimmer) [inline plugins/ltac/profile_ltac.ml:443]: What happened to this? Should the comment also be dropped?

File: lib/cWarnings.ml
Comment (SkySkimmer) [inline lib/cWarnings.ml:57]: This `current` stuff seems like it should be a different PR with changelog and doc.

Comment (gares) [reply lib/cWarnings.ml:57]: ok, I'll try to do that.

File: plugins/ltac/profile_ltac.ml
Comment (gares) [reply plugins/ltac/profile_ltac.ml:443]: This is part of a commit labelled "bug".

It is the only dependency of code above toplevel on the STM. This PR exists because I want to be able to not link the STM, so I had to remove it. I can do as I did for `par:` which would be that if the STM is not linked, you get a "degraded" feature. In this case it means that if you are using ltac profiling in interactive mode and you ump back and forth, the stats printed (eg at_exit) would be cumulative of all runs, not just of the "most recent one".

I'll make a separate PR as well if you prefer.

Comment (SkySkimmer) [reply plugins/ltac/profile_ltac.ml:443]: I don't know the ltac profiler enough to know how impactful that would be (are cumulative stats even a bad thing?)
As long as the comment is changed to avoid implying that checking is done I'm OK with it, maybe @coq/ltac-maintainers would like to comment.

**Comment by ejgallego** [comment]
@gares looks good to me, modulo some nits here and there that shouldn't block this.

Actually I remember now where I got blocked, and was indeed in the profile_ltac stuff, what I have been doing is either disabling this or just linking the STM in my toplevels. I think that handling that in a separate PR would be a good idea, so we can:
- merge this first, but still linking the stm, then remove the dependency later
- remove the dependency on the stm from the plugin, then merging this.

**Comment by gares** [comment]
about ltacprof, let's wait a word from @coq/ltac-maintainers to see if this change of behavior is tolerable as is of I should do as I did with `par:`.

Btw, I did split the commit about cWarnings to #13785 , you may want to comment there as well.

File: plugins/ltac/profile_ltac.ml
Comment (JasonGross) [reply plugins/ltac/profile_ltac.ml:443]: Cumulative stats are a bad thing, it means that profiling is not synchronous with the document state.  In coqtop interactive mode at least this should keep working.  I'm fine with other IDEs getting a degraded mode, but please emit a warning here when in this degraded mode.  Ideally it should be possible for other IDEs that support backtracking to have some way of doing/allowing/registering this synchronization

**Review by JasonGross — DISMISSED** [review]
Please keep the filtering of expired states in LtacProf, at the very least in coqtop

**Comment by gares** [comment]
OK, I'll add an API to let the document manager install a function to test if a state is valid.

**Comment by gares** [comment]
done! (still needs a rebase on top of #13785)

File: plugins/ltac/profile_ltac.ml
Comment (JasonGross) [inline plugins/ltac/profile_ltac.ml:439]: The indentation here seems weird, did the number of spaces change?

**Comment by JasonGross** [comment]
Great, thanks!

**Comment by gares** [comment]
Just rebased on top of master, I think it is finally ready.

File: dev/dune
Comment (SkySkimmer) [inline dev/dune:37]: The dev/db and dune_db* files should also be updated.

**Review by SkySkimmer — APPROVED** [review]

File: dev/dune
Comment (gares) [reply dev/dune:37]: I looked ad these files, but I don't know what to do (I never use Drop or ocamldebug...)

**Comment by gares** [comment]
thanks!

File: lib/stateid.mli
Comment (gares) [inline lib/stateid.mli:45]:
```suggestion
(* Asks the document manager if the given state is valid (or belongs to an
```

**Comment by gares** [comment]
I reviewed all the changes, which look good, but for a typo in a comment.
I think we should let CI finish, and then you can commit the fix just before merging the PR.

File: lib/stateid.mli
Comment (SkySkimmer) [reply lib/stateid.mli:45]: Isn't it your change in your PR?

Comment (gares) [reply lib/stateid.mli:45]: Yes, read the message below, or ignore this comment.

**Comment by SkySkimmer** [comment]
@coqbot: merge now

**Comment by gares** [comment]
Thanks!

**Pull request merged**
